Makefile

```diff
@@ -1,42 +1,74 @@
-###############################################################################
-# The build architecture is select by setting the ARCH variable.
-# # For example: When building on ppc64le you could use ARCH=ppc64le make <....>.
-# # When ARCH is undefined it defaults to amd64.
-ARCH?=amd64
-ifeq ($(ARCH),amd64)
-	ARCHTAG?=
+# Both native and cross architecture builds are supported.
+# The target architecture is select by setting the ARCH variable.
+# When ARCH is undefined it is set to the detected host architecture.
+# When ARCH differs from the host architecture a crossbuild will be performed.
+
+# restore this when all arches work. For now, there are qemu issues related to cross-compile on s390x
+#   see https://github.com/projectcalico/go-build/pull/32
+#ARCHES=$(patsubst docker-image/Dockerfile.%,%,$(wildcard docker-image/Dockerfile.*))
+ARCHES=amd64 arm64 ppc64le
+
+
+# BUILDARCH is the host architecture
+# ARCH is the target architecture
+# we need to keep track of them separately
+BUILDARCH ?= $(shell uname -m)
+BUILDOS ?= $(shell uname -s | tr A-Z a-z)
+
+# canonicalized names for host architecture
+ifeq ($(BUILDARCH),aarch64)
+        BUILDARCH=arm64
 endif
-
-ifeq ($(ARCH),ppc64le)
-	ARCHTAG:=-ppc64le
+ifeq ($(BUILDARCH),x86_64)
+        BUILDARCH=amd64
 endif
 
-ifeq ($(ARCH),s390x)
-	ARCHTAG:=-s390x
+# unless otherwise set, I am building for my own architecture, i.e. not cross-compiling
+ARCH ?= $(BUILDARCH)
+
+# canonicalized names for target architecture
+ifeq ($(ARCH),aarch64)
+        override ARCH=arm64
 endif
+ifeq ($(ARCH),x86_64)
+    override ARCH=amd64
+endif
+
+CONTAINER_NAME=calico/routereflector
+GO_BUILD_VER ?= v0.15
 
-.PHONEY: clean
+.PHONY: clean image
 
 # These variables can be overridden by setting an environment variable.
 LOCAL_IP_ENV?=$(shell ip route get 8.8.8.8 | head -1 | cut -d' ' -f8)
 
 CONFD_REPO?=calico/confd
 CONFD_VER?=v3.1.1
+# problem is that calico/conf does not yet have for other archs, and archless is amd64
+# so we will add a little logic here to handle it
+ifeq ($(ARCH),amd64)
 CONFD_CONTAINER_NAME=${CONFD_REPO}:${CONFD_VER}
+else
+CONFD_CONTAINER_NAME=${CONFD_REPO}:${CONFD_VER}-$(ARCH)
+endif
+
+default: clean image
 
-default: clean calicorr.created
+# standard build target, but nothing to do
+build:
 
-calicorr.created: $(BUILD_FILES) dist/confd
-	docker build $(DOCKER_EXTRA_BUILD_ARGS) -t calico/routereflector$(ARCHTAG) -f Dockerfile$(ARCHTAG) .
+image: calicorr.created
+calicorr.created: $(BUILD_FILES) dist/confd-$(ARCH)
```

Comment:
Probably need the arch in the name of the .created file or the first arch you build will prevent others from being built.

Reply:
Will add. Good point.

Reply:
I think I will make a variable out of it.